### PR TITLE
On calling /api/token/generate, Create identities/users in WIT

### PR DESCRIPTION
**Pull request by kishansagathiya**
To ensure that WIT can re-use the tokens, on doing /api/token/generate it creates identities/users in wit's db as well, as per auth's db
i.e.,
	- If identities and users are created in auth, they will be created in wit as well
	- If identities and users are updated in auth, they will be updated in wit as well
fixes issue #161 
       - If identities and users are already created, on /api/token/generate it would give a warning 
       ```
       unable to update user in WIT. Response status, unable to create user in WIT : unable to update user
       ```

### CODE REVIEW

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-1`

File: controller/token.go [hunk 1]
```diff
@@ -142,9 +143,29 @@ func (c *TokenController) Generate(ctx *app.GenerateTokenContext) error {
...
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "unable to persist user properly")
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to persist user properly")))
+	}
 	tokens = append(tokens, testuser)
 
+	var remoteWITService wit.RemoteWITServiceCaller
+	witURL, err := c.Configuration.GetWITURL(ctx.RequestData)
```
Comment (alexeykazakov) [inline controller/token.go]:
```
if err != nil {
	return jsonapi.JSONErrorResponse(ctx, err)
}
```

File: controller/token.go [hunk 2]
```diff
@@ -142,9 +143,29 @@ func (c *TokenController) Generate(ctx *app.GenerateTokenContext) error {
...
+	err = remoteWITService.CreateWITUser(ctx, ctx.RequestData, identity, witURL, identity.ID.String())
+	if err != nil {
+		log.Warn(ctx, map[string]interface{}{
+			"err":         err,
+			"identity_id": identity.ID,
+			"username":    identity.Username,
+			"wit_url":     witURL,
+		}, "unable to create user in WIT ")
+		//return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to create user in WIT ")))
```
Comment (alexeykazakov) [inline controller/token.go]: We don't need to keep this comment. So, let's remove it.

File: controller/token.go [hunk 3]
```diff
@@ -153,14 +174,27 @@ func (c *TokenController) Generate(ctx *app.GenerateTokenContext) error {
...
+	identity, _, err = c.Auth.CreateOrUpdateIdentity(ctx, *testuser.Token.AccessToken)
 	if err != nil {
 		log.Error(ctx, map[string]interface{}{
 			"err": err,
 		}, "unable to persist user properly")
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to persist user properly")))
 	}
 	tokens = append(tokens, testuser)
 
+	witURL, err = c.Configuration.GetWITURL(ctx.RequestData)
```
Comment (alexeykazakov) [inline controller/token.go]: Remove this line. You already got witURL.

**Comment by alexeykazakov** [comment]
[test]

**Comment by alexeykazakov** [comment]
[test]

**Comment by alexeykazakov** [comment]
[test]

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-2`

**Comment by alexeykazakov** [comment]
Fix the tests please.

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-3`

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-4`

**Comment by alexeykazakov** [comment]
Have you tested it? The Identity created in WIT has wrong ID for some reason (not the same as in Auth).

**Comment by kishansagathiya** [comment]
@alexeykazakov This was something that we had seen earlier, but for whatever reason after pulling latest changes, I am not seeing this problem anymore ( i.e identities/users created in both databases are same. Also identity_ids in users table matches with ids in identities table). I have checked this several times today and wasnt seeing any problem. Let me know if you are still facing this. 
Also, this mismatching problem will arise if we manually delete some entries from the database.

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-5`

**Comment by alexeykazakov** [comment]
@kishansagathiya right! I saw misleading identity_id in WIT logs. But it was identity_id extracted from the service account token (from context) by WIT logger. The created users and identities have correct IDs.

File: controller/token.go [hunk 4]
```diff
@@ -142,25 +143,60 @@ func (c *TokenController) Generate(ctx *app.GenerateTokenContext) error {
...
 		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to generate test token ")))
 	}
-	_, _, err = c.Auth.CreateOrUpdateIdentity(ctx, *testuser.Token.AccessToken)
+
+	identity, _, err := c.Auth.CreateOrUpdateIdentity(ctx, *testuser.Token.AccessToken)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "unable to persist user properly")
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to persist user properly")))
```
Comment (alexeykazakov) [inline controller/token.go]: This line is the reason of test failure. There is no public key available in the token manager used by tests to parse token properly. So the CreateOrUpdate function fails in tests.

Comment (alexeykazakov) [reply controller/token.go]: It's quite tricky to fix tests if we want to fails here. It requires some big refactoring in our tests. So, we could just ignore this error for now. Merge my PR to your topic branch for that - https://github.com/kishansagathiya/fabric8-auth/pull/1
I'm not authorized to push to your branch.

Comment (kishansagathiya) [reply controller/token.go]: @alexeykazakov Merged https://github.com/kishansagathiya/fabric8-auth/pull/1. Thanks a lot for this.

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-168-6`

**Comment by alexeykazakov** [comment]
[test]

**Pull request merged**